userguide/tutorials/payment_plugin.adoc

```diff
@@ -1,61 +1,119 @@
 
-When building PaymentTransactionInfoPlugin objects, it is very important to correctly populate the https://github.com/killbill/killbill-api/blob/master/src/main/java/org/killbill/billing/payment/plugin/api/PaymentTransactionInfoPlugin.java[status] type:
+The  `PaymentTransactionInfoPlugin` also has a `getProperties` method that returns a `PluginProperty` List. This can be used to return plugin specific properties to the client API 
 
-* if the payment is successful, the plugin should return PROCESSED
-* if the payment is rejected by the gateway (insufficient funds, fails AVS check, fraud detected, etc.), the plugin should return ERROR
-* if the payment requires a completion step (3D-S verification, HPP, etc.), the plugin should return PENDING
-* if the gateway wasn't contacted (DNS error, SSL handshake error, socket connect timeout, etc.), the plugin should return CANCELED
-* for all other cases (socket read timeout, 500 returned, etc.), the plugin should return UNDEFINED
+In addition, the `PaymentTransactionInfoPlugin` has a `getStatus` method that returns an  https://github.com/killbill/killbill-api/blob/4ae1c343a593de937415e21feecb9f5405037fa3/src/main/java/org/killbill/billing/payment/plugin/api/PaymentPluginStatus.java[PaymentPluginStatus].  It is very important to correctly populate the status type. The following table elaborates how the status should be populated:
 
-You should try to avoid UNDEFINED as much as possible, because it is the only case where Kill Bill cannot retry payments (since the payment may or may not have happened).
+|===
+|Status | Status Description
 
-==== Janitor integration
+|PROCESSED
+|Indicates that the payment is successful
+|ERROR
+|Indicates that the payment is rejected by the gateway (insufficient funds, fails AVS check, fraud detected, etc.)
+|PENDING
+|Indicates that the payment requires a completion step (3D-S verification, HPP, etc.)
+|CANCELED
+|Indicates that the gateway wasn't contacted (DNS error, SSL handshake error, socket connect timeout, etc.)
+|UNDEFINED
+|Should be used for all other cases (socket read timeout, 500 returned, etc.)
+|===
 
-Kill Bill will attempt to fix PENDING and UNKNOWN states by polling the plugin via `getPaymentInfo`: if the plugin subsequently returns PROCESSED for example (maybe by querying the gateway directly), the internal payment state (as well as invoice balance, etc.) will be reflected.
 
-The Janitor will match internal transactions against plugin transactions via the transaction id: make sure `getKbTransactionPaymentId` is correctly implemented.
+Kill Bill has a https://docs.killbill.io/latest/userguide_payment.html#_janitor[Janitor] system in place that attempts to fix *PENDING* and *UNKNOWN* states . It polls the plugin via `PaymentPluginApi#getPaymentInfo` method. If the plugin subsequently returns *PROCESSED*, the Janitor updates the internal payment state as well as invoice balance, etc.) accordingly
```

Comment:
For GoCardless, `getPaymentInfo` could be implemented using https://developer.gocardless.com/api-reference/#payments-get-a-single-payment for instance.

Reply:
@pierre , do you want the above to be added to the document?

Reply:
Yes, I think that would make the tutorial really _complete_ (as-in: the plugin would be useful to charge and retrieve details about payments). We should also add a `GET` example as a step 5 in the [README](https://github.com/killbill/killbill-gocardless-example-plugin#testing), something like:

```
curl -v \
     -u admin:password \
     -H "X-Killbill-ApiKey: bob" \
     -H "X-Killbill-ApiSecret: lazar" \
    'http://127.0.0.1:8080/1.0/kb/payments/<PAYMENT_ID>?withPluginInfo=true'
```

We should also explain in the tutorial what `withPluginInfo` does, i.e. when _true_, the info from the plugin `PaymentTransactionInfoPlugin` (including plugin properties) is returned.

Reply:
> Yes, I think that would make the tutorial really _complete_ (as-in: the plugin would be useful to charge and retrieve details about payments). We should also add a `GET` example as a step 5 in the [README](https://github.com/killbill/killbill-gocardless-example-plugin#testing), something like:

okay, in that case, I think we should write the code for the `getPaymentInfo` , that will help to explain it better. What do you think?

Reply:
Yup, agreed 👍  In the implementation, you could pass a few pieces of data from GoCardless into the plugin properties, to show that feature.